File: core/src/test/java/bop/kernel/VThreadTest.java
package bop.kernel;

import java.io.Closeable;
import java.text.DecimalFormat;
import java.util.concurrent.Executor;
import java.util.concurrent.atomic.AtomicLong;
import org.junit.jupiter.api.Test;

public class VThreadTest {
  static class MyCoolThread extends Thread {
    @Override
    public void run() {
      var current = Thread.currentThread();
      System.out.println(current.getClass().getName());
    }
  }

  @Test
  public void testThreadCurrent() throws Throwable {
    var thread = new MyCoolThread();
    thread.start();
    thread.join();
  }

  @Test
  public void benchSuspendResume() throws Throwable {
    final var counter = new AtomicLong();
    Thread.ofPlatform().start(() -> {
      final var opsPerSecFormat = new DecimalFormat("###,###,###");
      for (int i = 0; i < 100000; i++) {
        try {
          Thread.sleep(1000L);
        } catch (InterruptedException e) {
          throw new RuntimeException(e);
        }
        var count = counter.getAndSet(0L);
        if (count == 0L) {
          continue;
        }
        if (count < 1000) {
          continue;
        }
        var perOp = 1_000_000_000L / count - 5;
        System.out.println(perOp + " ns/op    " + opsPerSecFormat.format(count) + " ops/sec");
      }
    });
    var thread = VThread.start(new InlineExecutor(), (vthread) -> {
      while (true) {
        vthread.suspend();
        counter.incrementAndGet();
        //                System.out.println("after yield: " + Thread.currentThread() + "   state: "
        // + vthread.threadState() + "   carrier thread: " + vthread.carrierThread());
      }
    });

    for (int i = 0; i < 100_000_000; i++) {
      thread.resume();
      //      System.out.println("after continuation: " + Thread.currentThread() + "   state: " +
      // thread.threadState() + "    carrier thread: " + thread.carrierThread());
      //      Thread.sleep(1000L);
    }
  }

  private static class InlineExecutor implements Closeable, Executor {
    @Override
    public void execute(Runnable command) {
      command.run();
    }

    @Override
    public void close() {}
  }
}

File: jvm/core/src/test/java/bop/kernel/VThreadTest.kt
package bop.kernel

import org.junit.jupiter.api.Test
import java.io.Closeable
import java.text.DecimalFormat
import java.util.concurrent.Executor
import java.util.concurrent.atomic.AtomicLong
import java.util.function.Consumer

class VThreadTest {
   internal class MyCoolThread : Thread() {
      override fun run() {
         val current = currentThread()
         println(current.javaClass.getName())
      }
   }

   @Test
   fun testThreadCurrent() {
      val thread = MyCoolThread()
      thread.start()
      thread.join()
   }

   @Test
   fun benchSuspendResume() {
      val counter = AtomicLong()
      Thread.ofPlatform().start(Runnable {
         val opsPerSecFormat = DecimalFormat("###,###,###")
         for (i in 0..99999) {
            try {
               Thread.sleep(1000L)
            } catch (e: InterruptedException) {
               throw RuntimeException(e)
            }
            val count = counter.getAndSet(0L)
            if (count == 0L) {
               continue
            }
            if (count < 1000) {
               continue
            }
            val perOp = 1000000000L / count - 5
            println(perOp.toString() + " ns/op    " + opsPerSecFormat.format(count) + " ops/sec")
         }
      })
      val thread = VThread.start(InlineExecutor(), Consumer { vthread: VThread? ->
         while (true) {
            vthread!!.suspend()
            counter.incrementAndGet()
            //                System.out.println("after yield: " + Thread.currentThread() + "   state: "
            // + vthread.threadState() + "   carrier thread: " + vthread.carrierThread());
         }
      })

      for (i in 0..100000000 - 1) {
         thread.resume()
         //      System.out.println("after continuation: " + Thread.currentThread() + "   state: " +
         // thread.threadState() + "    carrier thread: " + thread.carrierThread());
         //      Thread.sleep(1000L);
      }
   }

   private class InlineExecutor : Closeable, Executor {
      override fun execute(command: Runnable) {
         command.run()
      }

      override fun close() {}
   }
}
